donjo-app/libraries/Release.php

```diff
@@ -0,0 +1,275 @@
+	 * 'Di rilis ini, versi 20.07, tersedia fitur untuk membuat file QR Code yg bisa dipasang di artikel,
+	 * surat atau materi lain. Rilis ini juga berisi perbaikan lain yang diminta Komunitas SID ... dst.'
+	 *
+	 * @return string
+	 */
+	public function getReleaseBody()
+	{
+		return $this->resyncWithOfficialRepository()->body;
+	}
+
+	/**
+	 * Sinkronisasi file cache dengan data di repositori.
+	 *
+	 * @return array
+	 */
+	public function resyncWithOfficialRepository()
+	{
+		if (! $this->api) {
+			throw new \Exception('Please specify the API endpoint URL.');
+		}
+
+		if ($this->cacheIsOutdated()) {
+			\Esyede\Curly::$certificate = FCPATH.'cacert.pem';
+
+			$options = array(CURLOPT_HTTPHEADER => array('Accept' => 'application/vnd.github.v3+json'));
+
+			$response = \Esyede\Curly::get($this->api, array(), $options);
+
+			if ($response instanceof \stdClass) {
+				$response = array(
+					'tag_name' => $response->body->tag_name,
+					'name' => $response->body->name,
+					'zipball_url' => $response->body->zipball_url,
+					'tarball_url' => $response->body->tarball_url,
+					'body' => $response->body->body,
+					'created_at' => $response->body->created_at,
+					'published_at' => $response->body->published_at,
+				);
+
+				$this->writeCache(json_encode($response));
+			}
+		}
+
+		return json_decode($this->readCache($this->cache));
+	}
+
+	/**
+	 * Cek apakah data cache sudah kadaluwarsa atau belum.
+	 * Cache dianggap kadaluwarsa jika:
+	 *   - File cachenya belum ada di server kita, atau
+	 *   - File cache sudah ada tetapi waktu modified-time filenya sudah
+	 *     lebih dari interval yang ditentukan.
+	 *
+	 * @return bool
+	 */
+	public function cacheIsOutdated()
+	{
+    return ! is_file($this->cache) || (time() > (filemtime($this->cache) + $this->interval));
+	}
+
+	/**
+	 * Ubah versi rilis menjadi integer agar bisa dibandingkan
+	 *
+	 * @param  string $version
+	 *
+	 * @return int
+	 */
+	public function fixVersioning($version)
+	{
+			$version = preg_replace('/[^0-9.]/', '', $version); // 'v20.07-pasca' -> '20.07'
+			$version = str_replace('.', '0', $version); // '20.07' -> '20007'
+			$version = (int) $version; // 20007 (integer)
+
+			return $version;
+		}
+
+	/**
+	 * Buat/timpa file cache jika sudah kadaluwarsa.
+	 *
+	 * @param  string $cache
+	 *
+	 * @return void
+	 */
+	public function writeCache($cache)
+	{
+		$file = $this->cache;
+		$interval = $this->interval;
+
+		if ($this->cacheIsOutdated()) {
+			if (is_file($file)) {
+				unlink($file);
+			}
+
+			file_put_contents($file, $cache, LOCK_EX);
+		}
+	}
+
+	/**
+	 * Baca file cache.
+	 *
+	 * @return string
+	 */
+	public function readCache()
+	{
+		return file_get_contents($this->cache);
+	}
+
+	/**
+	 * Cek apakah ada koneksi internet atau tidak.
+	 *
+	 * @return bool
+	 */
+	public function hasInternetConnection()
+	{
+		$connected = @fsockopen('www.google.com', 443);
+
+    if ($connected) {
+        fclose($connected);
+
+        return true;
+    }
```

Comment:
Ada helper serupa di `donjo-app/helpers/opensid_helper.php`. Mohon hindari duplikasi, kalau bisa. Kalau perlu perbaiki helper yg sudah ada. 

```
/**
 * Cek ada koneksi internet.
 *
 * @param            string $sCheckHost Default: www.google.com
 * @return           boolean
 */
function cek_koneksi_internet($sCheckHost = 'www.google.com')
{
	return (bool) @fsockopen($sCheckHost, 80, $iErrno, $sErrStr, 5);
}
```

Reply:
Resolved in a929fbe